### PR TITLE
ENH patch libgfortran for v4

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This PR needs to be merged before https://github.com/conda-forge/staged-recipes/pull/7288

### CODE REVIEW

**Comment by beckermr** [comment]
Good catch @isuruf! I have rewritten to make this more robust and to respect the minimum version if it is already past 3.0.0.

File: recipe/gen_patch_json.py [hunk 1]
```diff
@@ -239,6 +239,22 @@ def rename_dependency(fn, record, old_name, new_name):
...
+        if dep_idx:
+            # make sure respect minimum versions still there
+            # default to a min of 3.0.1 if none is given
+            if "3.0.1" in depends[dep_idx] or depends[dep_idx] == "libgfortran":
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.1,<4.0.1")
+            elif "4" in depends[dep_idx]:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=4.0.0,<4.0.1")
+            else:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.0,<4.0.1")
+
```
Comment (beckermr) [inline recipe/gen_patch_json.py]: Ack. We need to limit this only to osx...

Comment (beckermr) [reply recipe/gen_patch_json.py]: done.

File: recipe/gen_patch_json.py [hunk 2]
```diff
@@ -239,6 +239,22 @@ def rename_dependency(fn, record, old_name, new_name):
...
+        )
+        if dep_idx:
+            # make sure respect minimum versions still there
+            # default to a min of 3.0.1 if none is given
+            if "3.0.1" in depends[dep_idx] or depends[dep_idx] == "libgfortran":
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.1,<4.0.1")
+            elif "4" in depends[dep_idx]:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=4.0.0,<4.0.1")
+            else:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.0,<4.0.1")
```
Comment (isuruf) [inline recipe/gen_patch_json.py]: This should be changed to `<4.0.0.a0`

Comment (beckermr) [reply recipe/gen_patch_json.py]: OK I think I got this done.

File: recipe/gen_patch_json.py [hunk 3]
```diff
@@ -239,6 +239,22 @@ def rename_dependency(fn, record, old_name, new_name):
...
+    def fix_libgfortran(fn, record):
+        depends = record.get("depends", ())
+        dep_idx = next(
+            (q for q, dep in enumerate(depends) if dep.split(' ')[0] == "libgfortran"),
+            None
+        )
+        if dep_idx:
+            # make sure respect minimum versions still there
+            # default to a min of 3.0.1 if none is given
+            if ">=3.0.1" in depends[dep_idx] or depends[dep_idx] == "libgfortran":
```
Comment (isuruf) [inline recipe/gen_patch_json.py]: There's also `libgfortran >=3.0` that needs to be addressed.

Comment (beckermr) [reply recipe/gen_patch_json.py]: ack yes this patch is annoying...

Comment (beckermr) [reply recipe/gen_patch_json.py]: Actually no. I think `>=3.0` is caught by the last else case.

File: recipe/gen_patch_json.py [hunk 4]
```diff
@@ -239,6 +239,22 @@ def rename_dependency(fn, record, old_name, new_name):
...
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.1,<4.0.0.a0")
+            elif ">=4.0.0" in depends[dep_idx]:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=4.0.0,<5.0.0.a0")
+            else:
+                rename_dependency(fn, record, depends[dep_idx], "libgfortran >=3.0.0,<4.0.0.a0")
```
Comment (beckermr) [inline recipe/gen_patch_json.py]: This case should catch any other variants

Comment (isuruf) [reply recipe/gen_patch_json.py]:
```
'libgfortran',
 'libgfortran ==3.0.1',
 'libgfortran >=3.0',
 'libgfortran >=3.0.1'
```

Here are the 4 variants I see in conda-forge packages. I think we don't want to change a specific pin like `libgfortran ==3.0.1` here.

Comment (beckermr) [reply recipe/gen_patch_json.py]: Ahh ok. Will fix.

Comment (beckermr) [reply recipe/gen_patch_json.py]: Done.

File: recipe/gen_patch_json.py (hunk 3 above)
Comment (beckermr) [reply recipe/gen_patch_json.py]: Done.

**Comment by scopatz** [comment]
This looks good to me!

**Comment by beckermr** [comment]
Thank you @isuruf !

**Pull request merged**